File: 2024/03/main.R
#!/usr/bin/env -S Rscript --vanilla

read_input <- function(input_file) {
  input_file |> readr::read_file()
}

mul <- function(a, b) {
  a * b
}

part1 <- function(input) {
  input |>
    stringr::str_extract_all(
      stringr::regex("mul\\([:digit:]+,[:digit:]+\\)")
    ) |>
    unlist() |>
    tibble::as_tibble_col(column_name = "expr_str") |>
    dplyr::mutate(
      expr = purrr::map(expr_str, str2expression),
      res = unlist(purrr::map(expr, eval))
    ) |>
    dplyr::pull(res) |>
    sum()
}

part2 <- function(input) {
  multiplier <- 1
  do <- function() {
    multiplier <<- 1
    NA
  }
  dont <- function() {
    multiplier <<- 0
    NA
  }
  eval_dos <- function(expr) {
    eval(expr) * multiplier
  }

  input |>
    stringr::str_extract_all(
      stringr::regex("(mul\\([:digit:]+,[:digit:]+\\))|do\\(\\)|don't\\(\\)")
    ) |>
    unlist() |>
    tibble::as_tibble_col(column_name = "expr_str") |>
    dplyr::mutate(
      expr_str = purrr::modify_if(expr_str, \(x) x == "don't()", ~"dont()"),
      expr = purrr::map(expr_str, str2expression),
      res = unlist(purrr::map(expr, eval_dos))
    ) |>
    dplyr::pull(res) |>
    sum(na.rm = TRUE)
}

argv <- argparser::arg_parser(
  name = "main.R",
  description = paste(
    # UPDATE DAY
    "Advent of Code 2024, Day 3:",
    "Take input file containing raw text input, and",
    "print out solution to either part 1 or part 2."
  ),
  hide.opts = TRUE
) |>
  argparser::add_argument(
    arg = "input", help = "Input text file path."
  ) |>
  argparser::add_argument(
    arg = "part", help = "Which part to solve, either 1 or 2.",
    type = "integer"
  ) |>
  argparser::parse_args()

input <- argv$input |>
  read_input()

argv$part |>
  switch(input |> part1() |> cat(sep = "\n"),
    input |> part2() |> cat(sep = "\n")
  )

File: 2024/03/main.py
#!/usr/bin/env python3

import argparse
import re


def read_input(input_file: str) -> str:
    with open(input_file, "r") as f:
        input_str = "".join([line.strip() for line in f.readlines()])
    return input_str


def part1(input: str) -> int:
    matches = re.findall("(?<=mul\\()[0-9]+,[0-9]+(?=\\))", input)
    sum = 0
    for match in matches:
        a, b = (int(dig) for dig in match.split(","))
        sum += a * b
    return sum


def part2(input: str) -> int:
    matches = re.findall("do\\(\\)|don't\\(\\)|(?<=mul\\()[0-9]+,[0-9]+(?=\\))", input)
    print(matches)
    sum = 0
    go = True
    for match_ in matches:
        match match_:
            case "do()":
                go = True
            case "don't()":
                go = False
            case _:
                if go:
                    a, b = (int(factor) for factor in match_.split(","))
                    sum += a * b
    return sum


def main(input_file, part):

    input_str = read_input(input_file)

    match part:
        case 1:
            print(part1(input_str))
        case 2:
            print(part2(input_str))


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="""Advent of Code 2024, Day 3:
    Take input file containing raw text input, and
    print out solution to either part 1 or part 2."""
    )
    parser.add_argument("input", type=str, help="Input text file path.")
    parser.add_argument("part", type=int, help="Which part to solve, either 1 or 2.")
    args = parser.parse_args()

    if args.part > 2 or args.part < 1:
        parser.error("`part` argument must be 1 or 2.")

    main(args.input, args.part)
